#####pathway enrichment and annotation for our top hits
library(ggplot2)
library(GenomicRanges)
library(ChIPseeker)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
require(org.Hs.eg.db)
library(missMethyl)
library(bumphunter)
library(TissueEnrich)

results<-read.table("meta_BET_PREDO_ITU_random1.tbl",sep="\t",header=T)

####relation to Island => enriched in OpenSea
load("annot_epic.Rdata") #this is the EPIC annotation from Illumina
#background: all CpGs available in our study
annot<-merge(results,annot_epic,by.x=1,by.y="Name")
index<-which(annot$PvalueARE< 9e-8) #n=10,320

top<-table(annot$Relation_to_Island[index])/10320
#Island    N_Shelf    N_Shore    OpenSea    S_Shelf    S_Shore 
#0.07044574 0.03139535 0.09970930 0.68362403 0.03207364 0.08275194 

#dim(results)[1]-10320=747781
all<-table(annot$Relation_to_Island[-index])/747781
#Island    N_Shelf    N_Shore    OpenSea    S_Shelf    S_Shore 
#0.19177005 0.03582199 0.09761013 0.55802033 0.03319956 0.08357795 



#####Figure S3A#######

loc<-rbind(cbind(top, rep("significant CpGs",6)),cbind(all, rep("non-significant CpGs",6)))
loc<-as.data.frame(loc)
loc$Type<-rownames(loc)
colnames(loc)<-c("Percentage", "Cohort","Type")
loc$Percentage<-as.numeric(loc$Percentage)
loc$RoundedPercentage = round(loc$Percentage, 4)*100
loc$Label = paste(loc$RoundedPercentage,"%", sep="")
loc$Type<-gsub(".1","",loc$Type)
names(loc)[1]<-"Percentage_of_CpGs"
loc$Label[loc$Percentage_of_CpGs < 0.03] <-""
pdf("relation_to_Island_top_CpGs.pdf")
ggplot(loc, aes(x = factor(Cohort), y = Percentage_of_CpGs, fill = Type)) + 
geom_bar(position = position_stack(), stat = "identity", width = .7) +
geom_text(aes(label = Label), position = position_stack(vjust = 0.5), size = 4) + scale_fill_brewer(palette="RdBu")+
theme(axis.title = element_text(size = 14),axis.text = element_text(size = 13),legend.text=element_text(size=rel(1.3)))
dev.off() 

#more Open Sea? yes!
table(annot$Relation_to_Island[index])#7,055 Open Sea
table(annot$Relation_to_Island[-index])#41,7277  Open Sea
fisher.test(matrix(c(7055,10320-7055,417277,747781-417277),ncol=2,byrow=T))#OR=1.71,p=3.5393e-148

###map to genomic location: enriched for distal intergenic regions
index<-which(annot$PvalueARE<9e-08)#n=10,320
top<-annot[index,c("MarkerName","pos","chr")]
top$end<-top$pos+1
top4epic<-makeGRangesFromDataFrame(top,start.field ="pos",end.field="end",seqnames="chr",ignore.strand=T,keep.extra.columns=F)

all<-annot[-index,c("MarkerName","pos","chr")]
all$end<-all$pos+1
all4epic<-makeGRangesFromDataFrame(all,start.field ="pos",end.field="end",seqnames="chr",ignore.strand=T,keep.extra.columns=F)

all_anno<-annotatePeak(all4epic,TxDb=TxDb.Hsapiens.UCSC.hg19.knownGene,annoDb="org.Hs.eg.db",tssRegion = c(-500, 500))
all_anno
#Annotated peaks generated by ChIPseeker
#747781/747781  peaks were annotated
#Genomic Annotation Summary:
#  Feature  Frequency
#9           Promoter 23.2211035
#4             5' UTR  0.7776341
#3             3' UTR  2.9423053
#1           1st Exon  0.7629239
#7         Other Exon  5.2674513
#2         1st Intron 13.9976811
#8       Other Intron 23.0921888
#6 Downstream (<=300)  0.1401480
#5  Distal Intergenic 29.7985640

all_genomic<-c(23.22,0.78,2.94,0.76,5.27,14.00,23.09,0.14,29.80)
all_genomic<-as.data.frame(all_genomic)
all_genomic$position<-c("Promoter","5' UTR","3' UTR", "1st Exon","Other Exon","1st Intron","Other Intron",
                        "Downstream (<=3kb)","Distal Intergenic")


top_anno<-annotatePeak(top4epic,TxDb=TxDb.Hsapiens.UCSC.hg19.knownGene,annoDb="org.Hs.eg.db",tssRegion = c(-500, 500))
top_anno
Annotated peaks generated by ChIPseeker
#10320/10320  peaks were annotated
#Genomic Annotation Summary:
#  Feature   Frequency
#9           Promoter 13.10077519
#4             5' UTR  0.69767442
#3             3' UTR  2.50000000
#1           1st Exon  0.68798450
#7         Other Exon  4.00193798
#2         1st Intron 15.92054264
#8       Other Intron 23.60465116
#6 Downstream (<=300)  0.07751938
#5  Distal Intergenic 39.40891473


########Figure S3B#######
top_genomic<-c(13.10,0.70,2.50,0.69,4.00,15.92,23.60,0.08,39.41)
top_genomic<-as.data.frame(top_genomic)
top_genomic$position<-c("Promoter","5' UTR","3' UTR", "1st Exon","Other Exon","1st Intron","Other Intron",
                        "Downstream (<=3kb)","Distal Intergenic")


loc_genomic<-rbind(cbind(as.numeric(all_genomic[,1]), rep("non-significant CpGs",9),all_genomic[,2]),
                   cbind(as.numeric(top_genomic[,1]), rep("significant CpGs",9),top_genomic[,2]))
loc_genomic<-as.data.frame(loc_genomic)
colnames(loc_genomic)<-c("Percentage", "Cohort","Type")
loc_genomic$Percentage<-as.numeric(as.character(loc_genomic$Percentage))
loc_genomic$Label = paste(loc_genomic$Percentage,"%", sep="")
loc_genomic$Label[loc_genomic$Percentage <3] <-""
loc_genomic$Label[6]<-"14.00%"  #two digits to match other labels
loc_genomic$Label[9]<-"29.80%"  #two digits to match other labels
loc_genomic$Label[10]<-"13.10%"  #two digits to match other labels
loc_genomic$Label[14]<-"4.00%"  #two digits to match other labels
loc_genomic$Label[16]<-"23.60%"  #two digits to match other labels

names(loc_genomic)[1]<-"Percentage_of_CpGs"
pdf("location.pdf")
ggplot(loc_genomic, aes(x = factor(Cohort), y = Percentage_of_CpGs, fill = Type)) + 
  geom_bar(position = position_stack(), stat = "identity", width = .7) +
  geom_text(aes(label = Label), position = position_stack(vjust = 0.5), size = 4) + scale_fill_brewer(palette="RdBu") +
  theme(axis.title = element_text(size = 14),axis.text = element_text(size = 13),legend.text=element_text(size=rel(1.3)))
dev.off()#Figure S2B

#top CpGs more distal intergenic?: yes!
length(which(all_anno@detailGenomicAnnotation[,9])) #222828
length(which(top_anno@detailGenomicAnnotation[,9])) #4067
fisher.test(matrix(c(4067,10320-4067,222828,747781-222828),ncol=2,byrow=T))#OR=1.53,p=9.737055e-95

####pathway enrichments####
#mapping CpGs to nearest genes
genes <- annotateTranscripts(TxDb.Hsapiens.UCSC.hg19.knownGene)

#map significant genes
load("genes_CpGs_EPIC.Rdata") #epic_genes, created with ChipSeeker
index<-which(results$PvalueARE<9e-08) #n=10,320
top<-results[index,] 
index<-which(epic_genes$CpG %in% top$MarkerName) #n=10,320
write.table(epic_genes[index,],"significant_genes_all.txt",sep="\t",quote=F,row.names=F)
sig<-unique(epic_genes$name[index])#n=3,817
write.table(sig,"3817_unique_significant_genes.txt",sep="\t",quote=F,row.names=F,col.names=F)

#map all genes
#background genes: all CpGs which are in meta-analysis 
index<-which(epic_genes$CpG %in% results$MarkerName) #n=758,101 
all_genes<-epic_genes[index,]
write.table(all_genes,"background_genes_all.txt",sep="\t",quote=F,row.names=F)
genes<-unique(all_genes$name)#n=20,214
write.table(genes,"20214_unique_background_genes.txt",sep="\t",quote=F,row.names=F,col.names=F)
length(which(sig %in% genes)) #n=3,817, ok

#run FUMA for enrichment using the online tool
#GO terms
sets<-read.table("FUMA_3817_20214_top_hits/GS.txt",sep="\t",header=T)
go_bp<-sets[sets$Category=="GO_bp",]
go_cc<-sets[sets$Category=="GO_cc",]
go_mf<-sets[sets$Category=="GO_mf",]

go_bp$GeneSet<- factor(go_bp$GeneSet, levels = go_bp$GeneSet[order(-log10(go_bp$p))])
go_bp<-go_bp[order(go_bp$p),]
go_bp$ratio<-go_bp$N_overlap/go_bp$N_genes
#336 at FRD 0.05
write.table(go_bp[,c(1,2,3,4,5,6)],"go_bp_fdr.txt",sep="\t",quote=F,row.names=F) #####Table S4

go_cc$GeneSet<- factor(go_cc$GeneSet, levels = go_cc$GeneSet[order(-log10(go_cc$p))])
go_cc<-go_cc[order(go_cc$p),]
go_cc$ratio<-go_cc$N_overlap/go_cc$N_genes
#41 at FRD 0.05
write.table(go_cc[,c(1,2,3,4,5,6)],"go_cc_fdr.txt",sep="\t",quote=F,row.names=F) #####Table S5

go_mf$GeneSet<- factor(go_mf$GeneSet, levels = go_mf$GeneSet[order(-log10(go_mf$p))])
go_mf<-go_mf[order(go_mf$p),]
go_mf$ratio<-go_mf$N_overlap/go_mf$N_genes
#66 at FRD 0.05
write.table(go_mf[,c(1,2,3,4,5,6)],"go_mf_fdr.txt",sep="\t",quote=F,row.names=F) ######Table S6

#TF
index<-which(sets$Category=="TF_targets")
tf<-sets[index,]
tf$GeneSet<- factor(tf$GeneSet, levels = tf$GeneSet[order(-log10(tf$p))])
tf<-tf[order(tf$p),]
tf$ratio<-tf$N_overlap/tf$N_genes
#264 at FRD 0.05
write.table(tf[,c(1,2,3,4,5,6)],"tf_fdr.txt",sep="\t",quote=F,row.names=F) ######Table S7

##tissue enrichment
#GTEX specific
gtex<-read.table("FUMA_3817_20214_top_hits/gtex_v8_ts_DEG.txt",sep="\t",header=T)
gtex<-gtex[order(gtex$p),]
gtex$ratio<-gtex$N_overlap/gtex$N_genes
index<-which(gtex$adjP<0.05 & gtex$Category=="DEG.up") #n=46
gtex[index,c("GeneSet","p")] #brain included, top hit cortex with p=7.079794e-40
write.table(gtex[index,c(1,2,3,4,5,6)],"gtex_fdr.txt",sep="\t",quote=F,row.names=F) ######Table S8


#HPA as GTEX has no placenta available
gs<-GeneSet(geneIds=sig,organism='Homo Sapiens',
            geneIdType=SymbolIdentifier())
background<-GeneSet(geneIds=genes,organism='Homo Sapiens',
                    geneIdType=SymbolIdentifier())
output<-teEnrichment(gs,rnaSeqDataset = 1, backgroundGenes=background) #rnaSeqDataset=1 is human protein atlas

#from Vignette, default: tissueSpecificGeneType = 1, all genes
seEnrichmentOutput<-output[[1]]
enrichmentOutput<-setNames(data.frame(assay(seEnrichmentOutput),
                                      row.names = rowData(seEnrichmentOutput)[,1]),
                           colData(seEnrichmentOutput)[,1])
enrichmentOutput$Tissue<-row.names(enrichmentOutput)
#y=-log10(x) => convert to x via 10^y
enrichmentOutput$p<-10^(-enrichmentOutput$Log10PValue)
enrichmentOutput$FDR<-p.adjust(enrichmentOutput$p,method="BH")
which(enrichmentOutput$FDR<0.05) #cerebral cortex, skin


#####Figure 2A########
pdf("enrichment_tissues_HPR_all_genes.pdf")
ggplot(enrichmentOutput,aes(x=reorder(Tissue,-Log10PValue),y=Log10PValue,
                            label = Tissue.Specific.Genes,fill = Tissue))+
  geom_bar(stat = 'identity')+
  labs(x='', y = '-LOG10(P-Value)')+
  theme_bw()+
  theme(legend.position='none')+
  theme(plot.title = element_text(hjust = 0.5,size = 11),axis.title =
          element_text(size=10))+
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        panel.grid.major= element_blank(),panel.grid.minor = element_blank())+
  geom_hline(yintercept=-log10(0.003), linetype="dashed", 
             color = "black", linewidth=0.3)
dev.off()
